components/script/dom/compressionstream.rs

```diff
@@ -0,0 +1,276 @@
+            },
+        }
+    }
+
+    fn get_ref(&self) -> &Vec<u8> {
+        match self {
+            Compressor::Deflate(zlib_encoder) => zlib_encoder.get_ref(),
+            Compressor::DeflateRaw(deflate_encoder) => deflate_encoder.get_ref(),
+            Compressor::Gzip(gz_encoder) => gz_encoder.get_ref(),
+        }
+    }
+
+    fn get_mut(&mut self) -> &mut Vec<u8> {
+        match self {
+            Compressor::Deflate(zlib_encoder) => zlib_encoder.get_mut(),
+            Compressor::DeflateRaw(deflate_encoder) => deflate_encoder.get_mut(),
+            Compressor::Gzip(gz_encoder) => gz_encoder.get_mut(),
+        }
+    }
+
+    fn write_all(&mut self, buf: &[u8]) -> Result<(), io::Error> {
+        match self {
+            Compressor::Deflate(zlib_encoder) => zlib_encoder.write_all(buf),
+            Compressor::DeflateRaw(deflate_encoder) => deflate_encoder.write_all(buf),
+            Compressor::Gzip(gz_encoder) => gz_encoder.write_all(buf),
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        match self {
+            Compressor::Deflate(zlib_encoder) => zlib_encoder.flush(),
+            Compressor::DeflateRaw(deflate_encoder) => deflate_encoder.flush(),
+            Compressor::Gzip(gz_encoder) => gz_encoder.flush(),
+        }
+    }
+
+    fn try_finish(&mut self) -> io::Result<()> {
+        match self {
+            Compressor::Deflate(zlib_encoder) => zlib_encoder.try_finish(),
+            Compressor::DeflateRaw(deflate_encoder) => deflate_encoder.try_finish(),
+            Compressor::Gzip(gz_encoder) => gz_encoder.try_finish(),
+        }
+    }
+}
+
+/// <https://compression.spec.whatwg.org/#compressionstream>
+#[dom_struct]
+pub(crate) struct CompressionStream {
+    reflector_: Reflector,
+
+    /// <https://streams.spec.whatwg.org/#generictransformstream>
+    transform: Dom<TransformStream>,
+
+    /// <https://compression.spec.whatwg.org/#compressionstream-format>
+    format: CompressionFormat,
+
+    // <https://compression.spec.whatwg.org/#compressionstream-context>
+    #[ignore_malloc_size_of = "defined in flate2"]
+    #[no_trace]
+    context: RefCell<Compressor>,
+}
+
+impl CompressionStream {
+    fn new_inherited(transform: &TransformStream, format: CompressionFormat) -> CompressionStream {
+        CompressionStream {
+            reflector_: Reflector::new(),
+            transform: Dom::from_ref(transform),
+            format,
+            context: RefCell::new(Compressor::new(format)),
+        }
+    }
+
+    /// <https://compression.spec.whatwg.org/#dom-compressionstream-compressionstream>
+    fn new_with_proto(
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        proto: Option<SafeHandleObject>,
+        format: CompressionFormat,
+        can_gc: CanGc,
+    ) -> Fallible<DomRoot<CompressionStream>> {
+        // Step 1. If format is unsupported in CompressionStream, then throw a TypeError.
```

Comment:
Nit: these spec comments usually refer to the `Constructor`. We typically implement that in the `Constructor` and keep `new_with_proto` small.